Added support for mobile sms

nishikeshKardak opened the pull request:
Using way2sms API we can send mobile sms. We need way2sms acoount.

nishikeshKardak committed:
add support for sms

nishikeshKardak committed:
Update main.py

nishikeshKardak committed:
Add support for sms

nishikeshKardak committed:
Delete send_sms.py

nishikeshKardak committed:
Create send_sms.py

jarifibrahim commented:
Hey @nishikeshKardak :wave: 
Thank you so much for sending the PR. It looks good. :smiley: 

Can you please add some information to the PR description (and readme.md file) about the way2sms API and how I can test your PR?

For instance, I would like to know
1. What changed I need to make to test the PR
2. How to get the way2sms API key (if there is one)
3. What configuration parameters I need to pass so that I get the sms feature works.

nishikeshKardak commented:
WAY2SMS provides good API for sending sms.
Both the apiKey and secretKey are available at way2sms user account.

![capture4654](https://user-images.githubusercontent.com/39300717/51896730-18a44080-23d3-11e9-90c0-f87ffedf2225.JPG)

nishikeshKardak commented:
I made prepare_sms() method in general.py file, it prepares message to send. In prepare_sms() method send_sms() gets called with message as a parameter. send_sms() takes message as argument and called sendPostRequest() method with apiKey, secretKey, phoneNo and message.

jarifibrahim commented on send_sms.py:
```diff
@@ -0,0 +1,31 @@
+import json
+import general
+
+URL = 'http://www.way2sms.com/api/v1/sendCampaign'
```
It would make sense to move this (and other configs like apikey, secret key) to the config file https://github.com/jarifibrahim/resultBot/blob/master/config.py

nishikeshKardak replied:
Hey @jarifibrahim 👋
Thanks for reveiwing PL
As per your requested changes, I did all required changes.

jarifibrahim commented on main.py:
```diff
@@ -47,22 +47,21 @@ def get_files(files, dl_file_name):
-    print '\n' + str(datetime.now().strftime('%Y-%m-%d %H:%M:%S'))
-    print '=' * 40
-    print "Fetching results related to " + FACULTY + " faculty"
+    print ('\n' + str(datetime.now().strftime('%d-%b-%Y %H:%M:%S')))
```
Thanks for fixing this.
`30-Jan-2019 12:17:02'` looks much better than `2019-01-30 12:19:30` :)

nishikeshKardak replied:
Thanks for appreciating 😀

jarifibrahim commented on send_sms.py:
```diff
@@ -0,0 +1,31 @@
+
+
+# get request
+def sendPostRequest(reqUrl, apiKey, secretKey, useType, phoneNo, senderId, textMessage):
```
Most of the params for this method should be pulled in from the config file. 
Can we do something similar to https://github.com/jarifibrahim/resultBot/blob/master/send_email.py#L7 ?

jarifibrahim replied:
Can you please change this to `send_post_request`? 
I've tried to follow the PEP8 naming conventions. See https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

jarifibrahim commented on send_sms.py:
```diff
@@ -0,0 +1,31 @@
+URL = 'http://www.way2sms.com/api/v1/sendCampaign'
+
+
+# get request
```
The comment seems to be off :)

jarifibrahim replied:
@nishikeshKardak clean up this one

nishikeshKardak replied:
Referencing this #9

jarifibrahim reviewed (changes requested):
Hey @nishikeshKardak most of the stuff looks good. I've added a few comments. Feel free to ask for help :)

jarifibrahim commented on general.py:
```diff
@@ -1,22 +1,23 @@
 
 from bs4 import BeautifulSoup
-from urlparse import urljoin
+from urllib.parse import urlparse
```
Can you undo this change, please?
The project is intended to be used with python 2.7 and python 2.7 doesn't have a `urllib.parse` package.

nishikeshKardak committed:
Undo changes

Undo "from urllib.parse import urlparse" to "from urlparse import urljoin". Python 2.x  doesn't have urllib.urlparse package.

nishikeshKardak committed:
move authentication credentials to config.py

nishikeshKardak committed:
Update sms authentication credentials

nishikeshKardak committed:
update code with PEP8 naming conventions

nishikeshKardak committed:
Update send_sms.py

nishikeshKardak commented:
Hey @jarifibrahim I resolve all the problems. Can you please review PL.

jarifibrahim commented on send_sms.py:
```diff
@@ -0,0 +1,29 @@
+    except Exception as e:
+        print(e)
+    finally :
+        print()
```
Can we remove this, please?

nishikeshKardak replied:
Sure we can remove this.

jarifibrahim commented on send_sms.py:
```diff
@@ -0,0 +1,29 @@
+        print(e)
+    finally :
+        print()
+        exit()
```
We might not want to exit if we couldn't send the SMS. For instance, I was running this script on an AWS instance 24x7. If the SMS failed, it would've stopped the program. This is what we would want.

nishikeshKardak replied:
Hey @jarifibrahim if the SMS failed, then it would have stopped the program or not?

nishikeshKardak replied:
At this moment, I can't understand what to do if SMS failed. Can you help me?
If SMS failed then I would need to have exit from program or  have to continue?

jarifibrahim replied:
No. We should not kill the program if the sms failed. Let's keep the programming running even if the sms failed to send.

nishikeshKardak replied:
Hey @jarifibrahim I did following changes. Can you review this?

```
def send_sms(message):
    try:
        # get response
        response = send_post_request(WAY2SMS_URL, SMS_API_KEY, SMS_SECRET_KEY, 'stage', PHONE_NO, 'WAYSMS', message )
        # print response if you want
        res =  response.text
        if 'success' in res:
            print('\033[92m' + 'Successfully sent the sms' + '\033[0m')
    except requests.exceptions.HTTPError as eh:
        print('\033[91m' + 'An HTTP error occurred. %s' %eh)
        print('\033[0m') 
    except requests.exceptions.ConnectionError as ec:
        print('\033[91m' + 'A Connection error occurred %s' %ec)
        print('\033[0m')
    except requests.exceptions.Timeout as et:
        print('\033[91m' + 'It is ConnectTimeout or ReadTimeout %s' %et)
        print('\033[0m')
    except Exception as e:
        print('\033[91m' + "Something went wrong. Failed to send SMS: %s" % e)
        print('\033[0m')
    finally :
         pass
```

jarifibrahim replied:
@nishikeshKardak I think we don't need to handle all the exceptions. Look at this example
```python
>>> def try_and_print(req, params):
...  try:
...   requests.post(req, params)
...  except Exception as e:
...   print("Failed to send SMS: %s" % e)
>>> try_and_print("foo", "bar")
Failed to send SMS: Invalid URL 'foo': No schema supplied. Perhaps you meant http://foo?
>>> try_and_print("http:://google.com", "bar")
Failed to send SMS: No connection adapters were found for 'http:://google.com/'
>>> try_and_print("http://google.com", "bar")
>>> try_and_print("http://localhost", "bar")
Failed to send SMS: HTTPConnectionPool(host='localhost', port=80): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f89424932d0>: Failed to establish a new connection: [Errno 111] Connection refused',))
```

So a simple
```python
except Exception as e:
    print("Failed to send SMS: %s " % e)
```
should be enough

jarifibrahim commented on send_sms.py:
```diff
@@ -0,0 +1,29 @@
+        # print response if you want
+        print (response.text)
+    except Exception as e:
+        print(e)
```
it would be nice to print a better message along with the exception.
How about 
```suggestion
        print("Failed to send SMS: %s" % e)
```

jarifibrahim commented on main.py:
```diff
@@ -72,10 +71,11 @@ def main():
-        print 'Preparing email to send new files'
+        print ('Preparing email to send new files and Preparing sms to notify')
         prepare_email(new_files)
+        prepare_sms(new_files)
```
Can you make this optional, please?
Add a flag to config `NOTIFY_VIA_SMS` with `false` as default.
Then we can do 
```python
if NOTIFY_VIA_SMS:
        prepare_sms(new_files)
```

We should do this for the emails also https://github.com/jarifibrahim/resultBot/issues/6

jarifibrahim commented on config.py:
```diff
@@ -21,6 +21,12 @@
 
+# sms authentication credentials
+URL = 'http://www.way2sms.com/api/v1/sendCampaign'
+APIKEY = '' # way2sms apiKey
```
PEP8 naming conventions please :)
```suggestion
WAY2SMS_API_KEY = '' # way2sms apiKey
```

jarifibrahim commented on config.py:
```diff
@@ -21,6 +21,12 @@
+# sms authentication credentials
+URL = 'http://www.way2sms.com/api/v1/sendCampaign'
+APIKEY = '' # way2sms apiKey
+SECRETKEY = '' # way2sms secretKey
```
```suggestion
WAY2SMS_SECRET_KEY = '' # way2sms secretKey
```

jarifibrahim commented on config.py:
```diff
@@ -21,6 +21,12 @@
 EMAIL_SMTP = 'smtp.gmail.com:587'
 
+# sms authentication credentials
+URL = 'http://www.way2sms.com/api/v1/sendCampaign'
```
```suggestion
WAY2SMS_URL = 'http://www.way2sms.com/api/v1/sendCampaign'
```

jarifibrahim commented on config.py:
```diff
@@ -21,6 +21,12 @@
+URL = 'http://www.way2sms.com/api/v1/sendCampaign'
+APIKEY = '' # way2sms apiKey
+SECRETKEY = '' # way2sms secretKey
+PHONENO = # receiver's valid phone number(indian) without country code
```
```suggestion
PHONE_NO = # receiver's valid phone number(indian) without country code
```

jarifibrahim reviewed (approved):
Almost ready to merge :tada:

jarifibrahim commented:
@nishikeshKardak Just to confirm, have you tried running the PR with way2sms configurations? Did you receive the SMS?

nishikeshKardak committed:
Add flags to NOTIFY_VIA_MAIL and NOTIFY_VIA_SMS

nishikeshKardak committed:
make send_email and end_sms optional

nishikeshKardak committed:
Print better message along with exception.

nishikeshKardak committed:
Handled exceptions

nishikeshKardak commented:
> @nishikeshKardak Just to confirm, have you tried running the PR with way2sms configurations? Did you receive the SMS?

Yes,  I received sms with way2sms configuration. That's why I sent the PL.

nishikeshKardak commented:
> @nishikeshKardak Just to confirm, have you tried running the PR with way2sms configurations? Did you receive the SMS?

Yes, I received sms with way2sms configuration.

nishikeshKardak commented:
Hey @jarifibrahim  I did some changes. Can you review this??

nishikeshKardak commented:
Hey @jarifibrahim Thanks for merging PL.
